markdown/inlinepatterns.py

```diff
@@ -274,6 +261,27 @@ def get_stash(m):
         return util.INLINE_PLACEHOLDER_RE.sub(get_stash, text)
 
 
+class LegacyPattern(_BasePattern):
```

Comment:
I have very serious objections to the changes to this file. We have API changes. Under no circumstances am I interested in making API changes for type checking. If that means it's not possible to move forward with mypy, then we don't move forward with mypy.

To be clear, this specific line of code is not any more or less relevant than any other. However, the changes are rather sweeping and it was not possible to easily highlight and comment on all relevant lines. I'm speaking about most of the changes to this file in general.

Reply:
There is no actual change to the API, all usages can remain the same.

But this is still indeed the worst change here. I spent a lot of time thinking of this but it's the best I came up with. We could also solicit ideas regarding this.

I'll comment further on this soon.

Reply:
Let me explain what the problem here is.

This is the general explanation:
https://stackoverflow.com/questions/6034662/python-method-overriding-does-signature-matter

For the particular situation at hand:
`class InlineProcessor(Pattern):` violates the Liskov Substitution Principle.

And let's just quote directly from mypy's output:

```
Signature of "handleMatch" incompatible with supertype "Pattern"  [override]
     Superclass:
         def handleMatch(self, m: Match[str]) -> Element | str
     Subclass:
         def handleMatch(self, m: Match[str], data: str) -> tuple[Element | str | None, int | None, int | None]
```

The particular violation happens regardless of mypy, it's an actual run-time problem if users expect it to work like is widely accepted:

If you write:
```python
if isinstance(foo, Pattern):
    foo.handleMatch(single_argument)
```
it must work, because every instance of `Pattern` **must** have a single-argument `handleMatch` method, because it's written so right there!

An instance of `InlineProcessor` is an instance of `Pattern`, and every instance of it violates this principle.

More specific code example:

```pycon
>>> import re
>>> from markdown.inlinepatterns import EscapeInlineProcessor, Pattern

>>> foo = Pattern('a')
>>> foo.handleMatch(re.search('a', 'a'))  # OK

>>> foo = EscapeInlineProcessor('a')
>>> isinstance(foo, Pattern)
True
>>> foo.handleMatch(re.search('a', 'a'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: EscapeInlineProcessor.handleMatch() missing 1 required positional argument: 'data'
```



The correct hierarchy would be this:

```mermaid
  graph TD;
      BasePattern["BasePattern<br>defines helper methods only"]-->Pattern["Pattern<br>abstract def handleMatch(m):"];
      BasePattern-->InlineProcessor["InlineProcessor<br>abstract def handleMatch(m, data):"];
      Pattern-->SimpleTextPattern["SimpleTextPattern<br>def handleMatch(m):"]
      Pattern-->SimpleTagPattern["SimpleTagPattern<br>def handleMatch(m):"]
      InlineProcessor-->EscapeInlineProcessor["EscapeInlineProcessor<br>def handleMatch(m, data):"]
      InlineProcessor-->AsteriskProcessor["AsteriskProcessor<br>def handleMatch(m, data):"]
```

But the current hierarchy is:

```mermaid
  graph TD;
      Pattern["Pattern<br>defines helper methods<br>abstract def handleMatch(m):"];
      Pattern-->InlineProcessor["InlineProcessor<br>abstract def handleMatch(m, data):"];
      Pattern-->SimpleTextPattern["SimpleTextPattern<br>def handleMatch(m):"]
      Pattern-->SimpleTagPattern["SimpleTagPattern<br>def handleMatch(m):"]
      InlineProcessor-->EscapeInlineProcessor["EscapeInlineProcessor<br>def handleMatch(m, data):"]
      InlineProcessor-->AsteriskProcessor["AsteriskProcessor<br>def handleMatch(m, data):"]
```

As such, the entire hierarchy of `InlineProcessor` is said to have improper overloads.

What my change accomplishes is to construct the correct type hierarchy to mypy and to all typechecking users going forward.
**However at run time I am keeping everything just like it was. `LegacyPattern` is just a new alias to `Pattern`**

Reply:
See a code search of occurrences in the wild: currently every subclass of InlinePattern needs to add an ignore comment for this override.

https://github.com/search?q=%2FInlineProcessor%5C%29%2F+%2FhandleMatch.%2Bignore%2F+NOT+path%3Azerver&type=code

Of course at the moment that's not this repository's fault (as it doesn't have `py.typed`) but just because the type stubs are written in a way that directly mirrors it:
https://github.com/python/typeshed/blob/9ff393558a6f4fedc086bc9dc4386dbe529a3331/stubs/Markdown/markdown/inlinepatterns.pyi#L48

Reply:
There's also this issue report (which I found through the above code search because it's directly inspired by this case). Which if accepted would resolve this

* https://github.com/python/mypy/issues/10197

Reply:
An equivalent change to the external stubs would be this:
https://github.com/oprypin/typeshed/commit/43593a451a181726482dcfcce681e3f1d8b1bf08
Maybe this more condensed view is clearer to understand.